### ISSUE TITLE
Support sparse glyphs in .glyphs files

### ISSUE THREAD

**Issue by rsheeter** [googlefonts/fontc#62]
We see layer identifiers in Texturina that don't resolve. When this occurs it appears there is an associated master id:

```
associatedMasterId = "C4D96FDC-31F2-4CF4-BEA5-4CCE99FECD99";
layerId = "1FFA553E-9323-49A1-BA35-824C5E633887";
```

Not quite sure what this is; a "virtual layer"? Could also be associatedMasterId is unrelated, it just _sounds_ related and sits right beside layer ids :)

**Comment by rsheeter** [comment]
See https://github.com/Omnibus-Type/Texturina, sources/Texturina.glyphs

**Comment by anthrotype** [comment]
https://github.com/schriftgestalt/GlyphsSDK/blob/Glyphs3/GlyphsFileFormat/GlyphsFileFormatv3.md

> `associatedMasterId` string: ID of the master the layer is linked to. Not present if it equals layerID, i.e. if the layer is in use as master.

every layer in Glyphs.app is either a 'master' layer (one required master layer in each glyph, for each of the font masters) or it is a non-master layer, but it must still be associated with one of the master layers.
When associatedMasterId == layerId, then it is a master layer, and associatedMasterId is omitted on serializing.

Non master layers whose associatedMasterId != layerId can be either temporary drafts that are not taken into account when building the font, or they can be so-called "brace" or "bracket" layers... 

It's very complicated, and I don't have the time right now to fully explain. You could start by searching for "associatedMasterId" inside glyphsLib to get an idea.

How about you skip these for the time being, and we'll get back to this later?

**Comment by anthrotype** [comment]
> When associatedMasterId == layerId, then it is a master layer

to be more precise, a glyph layer corresponds to a font master when its layerId == master.id; it is also true that when this is the case, then its associatedMasterId == layerId (== master.id)

**Comment by rsheeter** [comment]
Per IM, to just compile the base default glyphs we can just ignore the ones whose layer id doesn't match any master. We'll use them for other things, perhaps alternates, but not the default glyph.

**Comment by anthrotype** [comment]
Non-master layers (associatedMasterId != layerId; layerId not in set(m.id for m in font.masters)) can be of several types. If you right click on such a layer in Glyphs.app you'll see from the drop-down menu the various kinds:

![image](https://user-images.githubusercontent.com/6939968/208107464-96c3f8b7-2998-4f6d-b5db-3894f5f6ec3e.png)

1) so-called "brace" layers, or additional intermediate masters for individual glyphs.
   For Glyphs v2, there's this doc: https://glyphsapp.com/learn/additional-masters-for-individual-glyphs-the-brace-trick
   For Glyphs v3, I haven't found an equivalent document yet. By looking at glyphsLib code, the difference seems to be that, instead of embedding the axis values directly in the layer.name like v2 does (e.g. "{500, 100}"), in v3 brace layers have a "coordinates" key in the layer.attributes containig the list of axis values, and the layer.name itself is irrelevant. I think 
   In this context, the associatedMasterId that links the given brace layer with some master, is used in order to fill in the omitted axis values with those from the associated master. E.g. if there are 3 axes but the brace layer only differ for the first one, its coordinates can contain only the first axis value and the rest are filled with the master's axis values.
   For reference, [here](https://github.com/googlefonts/glyphsLib/blob/6763648a91ecadee256d0518c731e35eb311c013/Lib/glyphsLib/builder/sources.py#L138-L209) is where glyhphsLib converts from brace layers to designspace source layers (stored as additional font layers named after the brace layer within the UFO of the associated master). And [here](https://github.com/googlefonts/glyphsLib/blob/6763648a91ecadee256d0518c731e35eb311c013/Lib/glyphsLib/classes.py#L3833-L3849) is how glyphsLib identifies a layer as "brace" in v2 and v3 and retrieves its coordinates.
   I think we want to support these as they enable a degree of sparseness in the variable glyph sources. Basically, speaking in terms of our IR, some variable glyphs may define more sources than others. In ufo2fontir we already support this by loading glifs from the given DS source.layer optional attribute (corresponding to the name of the UFO font layer where glyph sources are contained).

1) so-called "bracket" layers, i.e. alternate layers substituted when axis conditions are met (GSUB feature-variations)
    For Glyphs v3: https://glyphsapp.com/learn/switching-shapes
    For Glyphs v2: https://glyphsapp.com/learn/alternating-glyph-shapes and https://glyphsapp.com/learn/alternating-glyph-shapes-with-multiple-axes
    These are more complicated in that, even if Glyphs.app represent them as layers within a glyph, they actually end up becoming alternate _glyphs_ to be substituted via GSUB in the final font. In fact when doing glyphs=>UFO that's what we do, we create new glyphs from them and add designspace `<rules>` that define when to substitute which. I suggest we exclude these for the time being.

I would also ignore for now the various color layers. I presume "Color" means COLRv1, "Color Palette" means COLRv0, "iColor" means sbix (I know, Georg is good at naming stuff) and "svg" well is self-explanatory.

By the way, the type of layers are not mutually exclusive and one can define at the same time an intermediate (brace) layer that is also, say, an alternate (bracket) layer, and in fact that drop-down menu allows you to select multiple ticks at the same time. Actually my minds blows away trying to imagine such a hydra-headed layer...

**Comment by anthrotype** [comment]
Forgot to add, if a non-master layer is neither of these special layer types (e.g. immediately after you click "+" to create one), then it is not going to be compiled in the final font, but is simply used as a sort of draft, or maybe kept in the source as an earlier revision for whatever reason. We can safely ignore those.

**Comment by anthrotype** [comment]
I just heard there's also a new kind of masters in town, called "Virtual Master", see https://github.com/googlefonts/glyphsLib/issues/859

**Comment by khaledhosny** [comment]
Not exactly new, I think G2 had them.

**Comment by khaledhosny** [comment]
That is probably what is called sparse masters elsewhere.

**Comment by anthrotype** [comment]
I've come across the "sparse masters" terminology when talking about "brace" or intermediate layers, which work like additional masters that are only defined for specific glyphs (as opposed to master layers which by definition are defined for all the glyphs). E.g. the crossbar of lowercase e may need an additional master along weight axis.

These "Virtual Masters" are somewhat similar and in fact use the same mechanism used for intermediate or brace layers, as described in the "Optional: add a virtual master" paragraph of the https://glyphsapp.com/learn/creating-a-variable-font doc page. 

Below is the inline documentation provided when you add such a custom parameter from within Glyphs.app UI itself (emphasis mine):

> Defines a font master for a variable font, thereby extending its design space accordingly. Rather than a ‘real’ master, which is defined in File > Font Info > Masters, a virtual master can only be drawn in the form of a Brace layer, i.e., a glyph layer with a name consisting of or ending in the comma-separated design space coordinates between curly braces, e.g., ‘{100, 300}’. A virtual master makes most sense for _design axes that only apply to a limited number of glyphs_, e.g., an axis that controls the middle crossbar heights of letters like A, E, F and H. The main advantages are that only affected glyphs need to be managed, and kerning does not need to be redone for an extra ‘real’ master.

Looks like it should be doable to translate these to fontmake-rs' glyph intermediate representation, where the sparseness is the default rather than the exception.

**Comment by khaledhosny** [comment]
I think the only difference is that brace layers have an axis with existing full masters, while virtual masters don’t require a full master (other than the default).

**Comment by rsheeter** [comment]
https://github.com/googlefonts/fontc/pull/403 fixes this for .designspace; an equivalent .glyphs could be used to test this.

**Comment by anthrotype** [comment]
We now support intermediate (brace) layers but still don't support "Virtual Masters", so let's keep this open.

**Comment by anthrotype** [comment]
looks like glyphsLib (and thus fontmake) don't support these Virtual Masters yet, https://github.com/googlefonts/glyphsLib/issues/859

**Comment by rsheeter** [comment]
So many layer types. It would be quite helpful to have a wee writeup, akin to https://github.com/googlefonts/fontc/blob/main/resources/text/units.md, on all these different types of layers, what htey are for, etc.

**Comment by anthrotype** [comment]
hm yeah.. but Virtual Master is not a _layer_ type strictly speaking, but a font-wide custom parameter that allows to define some additional "masters" that unlike real ones do not require _all_ the glyphs to have a corresponding "master" layer associated with them (which also entails a new whole kerning dict). I'm still trying to wrap my mind around them.

**Comment by anthrotype** [comment]
Fixed by #569
